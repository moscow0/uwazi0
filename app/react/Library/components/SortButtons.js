import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { wrapDispatch } from 'app/Multireducer';
import { actions } from 'react-redux-form';
import { t } from 'app/I18N';
import { Icon } from 'UI';
import { DropdownList } from 'app/Forms';
import { propertyTypes } from 'shared/propertyTypes';

const isSortableType = type =>
  [propertyTypes.text, propertyTypes.date, propertyTypes.numeric, propertyTypes.select].includes(
    type
  );

const getMetadataSorts = templates =>
  templates.toJS().reduce((sorts, template) => {
    template.properties.forEach(property => {
      const sortable =
        property.filter &&
        (isSortableType(property.type) ||
          (property.inherit && isSortableType(property.inherit.type)));

      if (sortable && !sorts.find(s => s.name === property.name)) {
        const sortString = `metadata.${property.name}${property.inherit ? '.inheritedValue' : ''}`;
        sorts.push({
          label: property.label,
          name: property.name,
          value: sortString,
          type: property.type,
          context: template._id,
        });
      }
    });
    return sorts;
  }, []);

class SortButtons extends Component {
  constructor(props) {
    super(props);
    this.changeOrder = this.changeOrder.bind(this);
  }

  changeOrder() {
    const { sort, treatAs } = this.props.search;
    this.sort(sort, treatAs, this.props.search.order === 'asc' ? 'desc' : 'asc');
  }

  sort(property, defaultTreatAs, selectedSort) {
    const treatAs = defaultTreatAs;
    let order = selectedSort;

    if (treatAs && !selectedSort) {
      order = treatAs === 'string' ? 'asc' : 'desc';
    }

    const sort = { sort: property, order, treatAs };

    this.props.merge(this.props.stateProperty, sort);

    // TEST!!!
    const filters = { ...this.props.search, ...sort, userSelectedSorting: true };
    // -------
    delete filters.treatAs;

    if (this.props.sortCallback) {
      this.props.sortCallback({ search: filters }, this.props.storeKey);
    }
  }

  validateSearch() {
    const { search } = this.props;
    const _search = { ...search };
    if (_search.sort === '_score' && !_search.searchTerm) {
      _search.sort = 'creationDate';
      _search.order = 'desc';
    }
    return _search;
  }

  render() {
    const { templates } = this.props;
    const search = this.validateSearch();
    const metadataSorts = getMetadataSorts(templates);
    const commonSorts = [
      { label: 'Title', value: 'title', type: 'text', context: 'System' },
      { label: 'Date added', value: 'creationDate', type: 'number', context: 'System' },
      { label: 'Date modified', value: 'editDate', type: 'number', context: 'System' },
    ];
    if (search.searchTerm) {
      commonSorts.push({
        label: 'Search relevance',
        value: '_score',
        type: 'number',
        context: 'System',
      });
    }
    const sortOptions = [...commonSorts, ...metadataSorts].map(option => ({
      ...option,
      label: t(option.context, option.label, undefined, false),
    }));

    return (
      <div className="sort-buttons">
        <DropdownList
          className="sort-dropdown"
          value={search.sort}
          data={sortOptions}
          valueField="value"
          textField="label"
          onChange={selected =>
            this.sort(
              selected.value,
              selected.type === 'text' || selected.type === 'select' ? 'string' : 'number'
            )
          }
        />
        <button
          type="button"
          disabled={search.sort === '_score'}
          className={`sorting-toggle ${search.sort === '_score' && 'disabled'}`}
          onClick={this.changeOrder}
        >
          <Icon
            icon={search.order === 'asc' && search.sort !== '_score' ? 'arrow-up' : 'arrow-down'}
          />
        </button>
      </div>
    );
  }
}

SortButtons.propTypes = {
  searchDocuments: PropTypes.func,
  stateProperty: PropTypes.string,
  search: PropTypes.object,
  templates: PropTypes.object,
  merge: PropTypes.func,
  sortCallback: PropTypes.func,
  storeKey: PropTypes.string,
};

function mapStateToProps(state, ownProps) {
  let { templates } = state;
  const stateProperty = ownProps.stateProperty
    ? ownProps.stateProperty
    : `${ownProps.storeKey}.search`;

  if (ownProps.selectedTemplates && ownProps.selectedTemplates.count()) {
    templates = templates.filter(i => ownProps.selectedTemplates.includes(i.get('_id')));
  }

  const search = stateProperty
    .split(/[.,/]/)
    .reduce(
      (memo, property) => (Object.keys(memo).indexOf(property) !== -1 ? memo[property] : null),
      state
    );
  return { stateProperty, search, templates };
}

function mapDispatchToProps(dispatch, props) {
  return bindActionCreators({ merge: actions.merge }, wrapDispatch(dispatch, props.storeKey));
}

export { SortButtons, mapStateToProps };
export default connect(mapStateToProps, mapDispatchToProps)(SortButtons);
