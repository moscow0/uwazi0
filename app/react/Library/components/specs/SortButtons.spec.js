import React from 'react';
import { shallow } from 'enzyme';
import { fromJS as immutable } from 'immutable';

import { SortButtons, mapStateToProps } from '../SortButtons';

describe('SortButtons', () => {
  let component;
  let instance;
  let props;

  const render = () => {
    component = shallow(<SortButtons {...props} />);
    instance = component.instance();
  };

  beforeEach(() => {
    props = {
      sortCallback: jasmine.createSpy('sortCallback'),
      merge: jasmine.createSpy('merge'),
      search: { order: 'desc', sort: 'title' },
      templates: immutable([
        {
          _id: 'id',
          properties: [
            {},
            { filter: true, name: 'date', label: 'date', type: 'date' },
            { filter: true, name: 'number', label: 'number', type: 'numeric' },
            { filter: true, name: 'my_select', label: 'my select', type: 'select' },
            { filter: true, name: 'sortable_name', label: 'sortableProperty', type: 'text' },
            {
              filter: true,
              name: 'inherited1',
              label: 'inherited1',
              type: 'relationship',
              inherit: {
                type: 'text',
              },
            },
            {
              filter: true,
              name: 'inherited2',
              label: 'inherited2',
              type: 'relationship',
              inherit: {
                type: 'multiselect',
              },
            },
          ],
        },
      ]),
      stateProperty: 'search',
      storeKey: 'library',
    };
  });

  describe('Sort options', () => {
    it('should use templates sortable properties as options (with asc and desc for each)', () => {
      render();
      expect(component).toMatchSnapshot();
    });

    describe('when multiple options have the same name', () => {
      it('should not duplicate the entry', () => {
        props.templates = immutable([
          {
            _id: 'id',
            properties: [
              {},
              {
                filter: true,
                name: 'sortable_name',
                label: 'sortableProperty',
                type: 'text',
                _id: '_id',
              },
            ],
          },
          {
            properties: [
              {
                filter: true,
                name: 'sortable_name',
                label: 'anotherLabel',
                type: 'text',
                _id: '_id',
              },
            ],
          },
        ]);
        render();
        expect(component.find('.sort-dropdown').props().data.length).toBe(4);
        expect(component.find('.sort-dropdown').props().data).toContainEqual({
          context: 'id',
          label: 'sortableProperty',
          name: 'sortable_name',
          type: 'text',
          value: 'metadata.sortable_name',
        });
      });
    });

    describe('when there is an active search term', () => {
      it('should display sort by search relevance option', () => {
        props.search.searchTerm = 'keyword';
        render();
        expect(component.find('.sort-dropdown').props().data).toContainEqual({
          context: 'System',
          label: 'Search relevance',
          type: 'number',
          value: '_score',
        });
      });

      it('should disable the sort button and show the sort down icon', () => {
        props.search.searchTerm = 'keyword';
        props.search.sort = '_score';
        render();
        expect(component.find('.sorting-toggle').props().disabled).toBe(true);
        expect(component.find('.sorting-toggle').props().children.props.icon).toBe('arrow-down');
      });
    });

    describe('when relevance sorting is active and there is no search term', () => {
      it('should fall back to sorting by most recent creation date', () => {
        props.search.sort = '_score';
        render();
        expect(component).toMatchSnapshot();
      });
    });

    describe('when active', () => {
      it('should set the option active', () => {
        props.search.sort = 'metadata.sortable_name';
        render();
        expect(component.find('.sort-dropdown').props().value).toBe('metadata.sortable_name');
      });
    });

    describe('clicking an option', () => {
      it('should sort by that property with default order (asc for text and desc for date)', () => {
        render();
        component
          .find('.sort-dropdown')
          .simulate('change', { value: 'metadata.sortable_name', type: 'select' });
        expect(props.sortCallback).toHaveBeenCalledWith(
          { search: { sort: 'metadata.sortable_name', order: 'asc', userSelectedSorting: true } },
          'library'
        );

        component
          .find('.sort-dropdown')
          .simulate('change', { value: 'metadata.different_name', type: 'date' });
        expect(props.sortCallback).toHaveBeenCalledWith(
          { search: { sort: 'metadata.different_name', order: 'desc', userSelectedSorting: true } },
          'library'
        );
      });
    });
  });

  describe('sort', () => {
    it('should not fail if no sortCallback', () => {
      delete props.sortCallback;
      render();
      let error;
      try {
        instance.sort('title');
      } catch (err) {
        error = err;
      }
      expect(error).toBeUndefined();
    });

    it('should change sorting order when using the order button', () => {
      props.search = { order: 'asc', sort: 'title' };
      render();
      component.find('button[type="button"]').simulate('click');
      expect(props.sortCallback).toHaveBeenCalledWith(
        { search: { sort: 'title', order: 'desc', userSelectedSorting: true } },
        'library'
      );
    });
    describe('when changing property being sorted', () => {
      it('should use default order', () => {
        props.search = { order: 'desc', sort: 'title' };
        render();
        instance.sort('title', 'string');
        expect(props.sortCallback).toHaveBeenCalledWith(
          { search: { sort: 'title', order: 'asc', userSelectedSorting: true } },
          'library'
        );

        props.sortCallback.calls.reset();
        props.search = { order: 'desc', sort: 'title' };
        render();
        instance.sort('creationDate', 'number');
        expect(props.sortCallback).toHaveBeenCalledWith(
          { search: { sort: 'creationDate', order: 'desc', userSelectedSorting: true } },
          'library'
        );

        props.sortCallback.calls.reset();
        props.search = { order: 'desc', sort: 'title' };
        render();
        instance.sort('creationDate', 'number');
        expect(props.sortCallback).toHaveBeenCalledWith(
          { search: { sort: 'creationDate', order: 'desc', userSelectedSorting: true } },
          'library'
        );
      });
    });

    describe('when changing order', () => {
      it('should keep the treatAs property', () => {
        props.search = { order: 'desc', sort: 'title', treatAs: 'number' };
        render();
        instance.sort('title', 'number');
        instance.changeOrder();
        expect(props.merge).toHaveBeenCalledWith('search', {
          sort: 'title',
          order: 'asc',
          treatAs: 'number',
        });
      });
    });
  });

  describe('when filtering number property', () => {
    it('should set value to number', () => {
      props.search = { order: 'asc', sort: 'number' };
      render();
      expect(component.find('.sort-dropdown').props().value).toBe('number');
    });
  });

  describe('mapStateToProps', () => {
    let templates;

    it('should send all templates from state', () => {
      const state = { templates: immutable(['item']), library: { search: {} } };
      const _props = { storeKey: 'library' };
      expect(mapStateToProps(state, _props).templates.get(0)).toBe('item');
    });

    it('should only send selectedTemplates if array passed in ownProps', () => {
      templates = immutable([{ _id: 'a' }, { _id: 'b' }]);
      const state = { templates, library: { search: {} } };
      const _props = { selectedTemplates: immutable(['b']), storeKey: 'library' };
      expect(mapStateToProps(state, _props).templates.getIn([0, '_id'])).toBe('b');
    });

    describe('search', () => {
      beforeEach(() => {
        templates = immutable([{ _id: 'a' }, { _id: 'b' }]);
      });

      it('should be selected from the state according to the store key', () => {
        const state = { templates, library: { search: 'correct search' } };
        const _props = { storeKey: 'library' };
        expect(mapStateToProps(state, _props).search).toBe('correct search');
      });

      it('should be selected from the state according to the stateProperty (if passed)', () => {
        let state = { templates, library: { search: 'incorrect search', sort: 'correct search' } };
        let _props = { storeKey: 'library', stateProperty: 'library.sort' };
        expect(mapStateToProps(state, _props).search).toBe('correct search');

        state = {
          templates,
          library: {
            search: 'incorrect search',
            sort: 'incorrect search',
            nested: { dashed: 'correct search' },
          },
        };
        _props = { storeKey: 'library', stateProperty: 'library/nested.dashed' };
        expect(mapStateToProps(state, _props).search).toBe('correct search');
      });
    });
  });
});
