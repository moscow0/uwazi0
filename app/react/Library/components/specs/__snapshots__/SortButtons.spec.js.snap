// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortButtons Sort options should use templates sortable properties as options (with asc and desc for each) 1`] = `
<div
  className="sort-buttons"
>
  <Uncontrolled(DropdownList)
    className="sort-dropdown"
    data={
      Array [
        Object {
          "context": "System",
          "label": "Title",
          "type": "text",
          "value": "title",
        },
        Object {
          "context": "System",
          "label": "Date added",
          "type": "number",
          "value": "creationDate",
        },
        Object {
          "context": "System",
          "label": "Date modified",
          "type": "number",
          "value": "editDate",
        },
        Object {
          "context": "id",
          "label": "date",
          "name": "date",
          "type": "date",
          "value": "metadata.date",
        },
        Object {
          "context": "id",
          "label": "number",
          "name": "number",
          "type": "numeric",
          "value": "metadata.number",
        },
        Object {
          "context": "id",
          "label": "my select",
          "name": "my_select",
          "type": "select",
          "value": "metadata.my_select",
        },
        Object {
          "context": "id",
          "label": "sortableProperty",
          "name": "sortable_name",
          "type": "text",
          "value": "metadata.sortable_name",
        },
        Object {
          "context": "id",
          "label": "inherited1",
          "name": "inherited1",
          "type": "relationship",
          "value": "metadata.inherited1.inheritedValue",
        },
      ]
    }
    onChange={[Function]}
    textField="label"
    value="title"
    valueField="value"
  />
  <button
    className="sorting-toggle false"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <Connect(Icon)
      icon="arrow-down"
    />
  </button>
</div>
`;

exports[`SortButtons Sort options when relevance sorting is active and there is no search term should fall back to sorting by most recent creation date 1`] = `
<div
  className="sort-buttons"
>
  <Uncontrolled(DropdownList)
    className="sort-dropdown"
    data={
      Array [
        Object {
          "context": "System",
          "label": "Title",
          "type": "text",
          "value": "title",
        },
        Object {
          "context": "System",
          "label": "Date added",
          "type": "number",
          "value": "creationDate",
        },
        Object {
          "context": "System",
          "label": "Date modified",
          "type": "number",
          "value": "editDate",
        },
        Object {
          "context": "id",
          "label": "date",
          "name": "date",
          "type": "date",
          "value": "metadata.date",
        },
        Object {
          "context": "id",
          "label": "number",
          "name": "number",
          "type": "numeric",
          "value": "metadata.number",
        },
        Object {
          "context": "id",
          "label": "my select",
          "name": "my_select",
          "type": "select",
          "value": "metadata.my_select",
        },
        Object {
          "context": "id",
          "label": "sortableProperty",
          "name": "sortable_name",
          "type": "text",
          "value": "metadata.sortable_name",
        },
        Object {
          "context": "id",
          "label": "inherited1",
          "name": "inherited1",
          "type": "relationship",
          "value": "metadata.inherited1.inheritedValue",
        },
      ]
    }
    onChange={[Function]}
    textField="label"
    value="creationDate"
    valueField="value"
  />
  <button
    className="sorting-toggle false"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <Connect(Icon)
      icon="arrow-down"
    />
  </button>
</div>
`;
